lib/Sema/ConstraintLocator.h

```diff
@@ -715,6 +715,21 @@ class ConstraintLocatorBuilder {
     return (getSummaryFlags() & ConstraintLocator::IsFunctionConversion);
   }
 
+  bool isAutoclosureResult() const {
+    SmallVector<LocatorPathElt, 4> path;
+    getLocatorParts(path);
+
+    auto last = std::find_if(
+        path.rbegin(), path.rend(), [](LocatorPathElt &elt) -> bool {
+          return elt.getKind() != ConstraintLocator::OptionalPayload;
```

Comment:
I think we'll also want to look through `ConstraintLocator::GenericArgument` as well in order to handle optional-to-optional conversions such as:

```swift
func foo(_ x: @autoclosure () -> UnsafePointer<Int>?) {}

var x: [Int]? = []
foo(x)
```

Reply:
This is going to be a source break as well just like `Int` to `UnsafePointer<Int>?` in one of the tests. We might need to get Core Team involved in this. /cc @DougGregor

Reply:
@hborla See https://github.com/apple/swift/pull/26543 that's one of the places where we'd like to add a new check for autoclosure.

Reply:
The Core Team is okay with accepting this source break given that this code never should have compiled in the first place is can trigger undefined behavior. Thanks for your patience.